apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ir/IRBuilder.kt

```diff
@@ -139,35 +139,34 @@ class IRBuilder(private val schema: IntrospectionSchema,
     }
 
     return try {
-      antlrParse(this, "").value().toKotlin()
+      antlrParse(this, "").value().toKotlin(type)
     } catch (e: Exception) {
       // seatbelt mode on, catch anything that can go wrong and use the previous behavour as fallback
       normalizeValue(type)
     }
   }
 
   /**
-   * This doesn't validate against the expected type, we're assuming the introspection schema is good
+   * Converts a GraphQLParser.ValueContext to its corresponding native type
    */
-  private fun GraphQLParser.ValueContext.toKotlin(): Any? {
+  private fun GraphQLParser.ValueContext.toKotlin(type: IntrospectionSchema.TypeRef): Any? {
     return when(this) {
       is GraphQLParser.StringValueContext -> this.STRING()?.text?.trimStart('\"')?.trimEnd('\"')
       is GraphQLParser.BooleanValueContext -> this.BOOLEAN()?.text?.toBoolean()
       is GraphQLParser.InlineInputTypeValueContext -> {
         this.inlineInputType().inlineInputTypeField().map {
-          it.NAME() to it.valueOrVariable()?.value()?.toKotlin()
+          it.NAME() to it.valueOrVariable()?.value()?.toKotlin(type)
         }.toMap()
       }
-      is GraphQLParser.NumberValueContext -> if (
-          NUMBER().text.contains('.') ||
-          NUMBER().text.contains("e")) {
-        NUMBER().text.toDouble()
-      } else {
-        NUMBER().text.toInt()
-      }
+      is GraphQLParser.NumberValueContext ->
+        when (ScalarType.forName(type.name ?: "")) {
+          ScalarType.INT -> NUMBER().text.toInt()
+          ScalarType.FLOAT -> NUMBER().text.toDouble()
+          else -> throw ParseException("Unexpected number type")
+        }
       is GraphQLParser.ArrayValueContext -> {
         this.arrayValueType().valueOrVariable()?.map {
-          it.value()?.toKotlin()
+          it.value()?.toKotlin(type)
```

Comment:
For lists, this should "unwrap" the item type with `type.ofType` as it recurses in order to coerce scalar items appropriately. For an example with `defaultValue = "[3]"`, `type.kind == LIST` and `type.ofType.kind == SCALAR`. Same thing for input fields where it gets even more complicated as we'll need a reference to the schema to lookup the inputFieldDefinition.